python_examples/spatial_example/voronoi_plotting_example.py

```diff
@@ -0,0 +1,231 @@
+    for r, region in enumerate(vor.regions):
+        if -1 not in region:
+            point = vor.points[r]
+            value = find_value_for_region(current_data, point, name=name)
+            polygon = [vor.vertices[i] for i in region]
+            ax.fill(*zip(*polygon), color=mapper.to_rgba(value))
+
+    ax.set_aspect('equal')
+    ax.set_title(f"t = {time:.2f}")
+    ax.set_xlim(0, 1)
+    ax.set_ylim(0, 1)
+    return fig, ax
+
+
+def plot_time_grid(df, vor, name, grid_dim, save_loc):
+    """Plots a grid of spatial plot of all cells in the Voronoi
+    tesselation, colour coded by their value in column 'name',
+    for multiple times.
+
+    :param df: Overall dataframe (from simulation output)
+    :type df: pd.Dataframe
+    :param vor: Voronoi tesselation object
+    :type vor: voronoi object
+    :param name: Name of quantity to extract
+    :type name: str
+    :param grid_dim: Size of grid of spatial plots
+    :type time: Tuple(int, int)
+    :param save_loc: Path of saved image
+    :type save_loc: str
+    """
+    # Generate colour map to use
+    mapper = generate_colour_map(df, name=name)
+
+    # Configure subplot grid
+    fig, axs = plt.subplots(grid_dim[0], grid_dim[1],
+                            sharex=True, sharey=True)
+    fig.subplots_adjust(hspace=0, wspace=.25)
+    axs = axs.ravel()
+
+    # Determine time points to use
+    plot_num = math.prod(grid_dim)
+    times = find_time_points(df["time"], plot_num)
+
+    # Add each subplot
+    for i, t in enumerate(times):
+        plot_time_point(df, vor, name, t, axs[i], mapper)
+
+    cbar = plt.colorbar(mapper, ax=axs.tolist())
+    cbar.set_label("Number of " + str(name))
+    plt.xlim((0, 1)), plt.ylim((0, 1))
+    plt.savefig(save_loc)
+
+
+def generate_animation(df, vor, name, save_path, use_pillow=True):
+    """Plots a grid of spatial plot of all cells in the Voronoi
+    tesselation, colour coded by their value in column 'name',
+    for numltiple times.
+
+    Has the option to use PillowWriter to generate the animation
+    on the fly, or generate the images separately and then compile
+    into an animation (recommended for older machines).
+
+
+    :param df: Overall dataframe (from simulation output)
+    :type df: pd.Dataframe
+    :param vor: Voronoi tesselation object
+    :type vor: voronoi object
+    :param name: Name of quantity to extract
+    :type name: str
+    :param save_path: Path to saved animation
+    :type save_path: str
+    :param use_pillow: Whether to use pillow to generate animations
+        on the fly, or from temporary images stored in memory
+    :type use_pillow: bool
+    """
+    # Generate colour map to use
+    mapper = generate_colour_map(df, name=name)
+
+    times = df["time"].to_numpy()
+
+    fig = plt.figure()
+    ax = plt.axes(xlim=(0, 1), ylim=(0, 1))
+
+    def animate(i):
+        return plot_time_point(df, vor, name, i, ax, mapper)
+
+    if use_pillow:
```

Comment:
@rcw5890  I thought about doing this too, but was aware PEP8 has strong opinions on the matter (i.e. never do that). A little exploration on stack exchange ([1](https://stackoverflow.com/questions/128478/should-import-statements-always-be-at-the-top-of-a-module), [2](https://stackoverflow.com/questions/1024049/is-it-pythonic-to-import-inside-functions)) seems to say that the community generally agrees, but there's certainly a range of opinions. 

Given the main reasons to use this construction appear to be when the module is very heavy to import, or when you might have dependency issues. Given neither of these are the case here (indeed all users will have matplotlib installed already), and its not inconceivable that a user might want to generate multiple animations in a single script, I personally think it is better to leave it how it is. Happy to be convinced otherwise though!

Reply:
> @rcw5890 Is this instead of imports at the top of the file? I thought about doing this too, but was aware PEP8 has strong opinions on the matter (i.e. never do that). A little exploration on stack exchange ([1](https://stackoverflow.com/questions/128478/should-import-statements-always-be-at-the-top-of-a-module), [2](https://stackoverflow.com/questions/1024049/is-it-pythonic-to-import-inside-functions)) seems to say that the community generally agrees, but there's certainly a range of opinions.
> 
> Given the main reasons to use this construction appear to be when the module is very heavy to import, or when you might have dependency issues. Given neither of these are the case here (indeed all users will have matplotlib installed already), and its not inconceivable that a user might want to generate multiple animations in a single script, I personally think it is better to leave it how it is. Happy to be convinced otherwise though!

Good point, I agree that there's nothing difficult about this optional import, so yes, we can put it at the top in the usual place.
I think we just need `import matplotlib.animation` somewhere since animation is not imported by default

Reply:
Ahh I see, I had resisted VSCode's suggestions to add that in because I already have matplotlib imported and it felt like I was importing every submodule separately too, I can add that though!